instancetype: Follow ups from #8537

lyarwood opened the pull request:
**What this PR does / why we need it**:

This PR addresses various clean ups and refactors identified in #8537.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #

**Special notes for your reviewer**:

**Release note**:
<!--  Write your release note:
1. Enter your extended release note in the below block. If the PR requires additional action from users switching to the new release, include the string "action required".
2. If no release note is required, just write "NONE".
-->
```release-note
NONE
```

lyarwood commented:
/cc @0xFelix follow ups as discussed.

lyarwood commented:
/cc @0xFelix

Follow ups as discussed.

lyarwood commented:
/area instancetype

0xFelix commented on pkg/instancetype/instancetype_test.go:
```diff
@@ -559,15 +539,13 @@ var _ = Describe("Instancetype and Preferences", func() {
 			It("returns expected preference using only the client", func() {
-				instancetypeMethods = instancetype.NewMethods(nil, nil, nil, nil, nil, virtClient)
-
+				instancetypeMethods = &instancetype.InstancetypeMethods{Clientset: virtClient}
```
This looks way better!

0xFelix reviewed (approved):
/lgtm

Great, thank you!

0xFelix commented:
/retest

lyarwood commented:
/hold

I'm going to use this PR to address the following lint feedback....

```
golangci-lint run pkg/instancetype/...
WARN [runner] The linter 'varcheck' is deprecated (since v1.49.0) due to: The owner seems to have abandoned the linter.  Replaced by unused. 
WARN [runner] The linter 'deadcode' is deprecated (since v1.49.0) due to: The owner seems to have abandoned the linter.  Replaced by unused. 
WARN [runner] The linter 'structcheck' is deprecated (since v1.49.0) due to: The owner seems to have abandoned the linter.  Replaced by unused. 
WARN [linters_context] rowserrcheck is disabled because of generics. You can track the evolution of the generics support by following the https://github.com/golangci/golangci-lint/issues/2649. 
WARN [linters_context] structcheck is disabled because of generics. You can track the evolution of the generics support by following the https://github.com/golangci/golangci-lint/issues/2649. 
pkg/instancetype/instancetype.go:416: 416-444 lines are duplicate of `pkg/instancetype/instancetype.go:571-599` (dupl)
func (m *InstancetypeMethods) findPreferenceSpecRevision(namespacedName types.NamespacedName) (*instancetypev1alpha2.VirtualMachinePreferenceSpec, error) {
        var (
                err      error
                revision *appsv1.ControllerRevision
        )

        if m.ControllerRevisionStore != nil {
                revision, err = m.getControllerRevisionByInformer(namespacedName)
        } else {
                revision, err = m.getControllerRevisionByClient(namespacedName)
        }

        if err != nil {
                return nil, err
        }

        if err = decodeRevisionObject(revision, true); err != nil {
                return nil, err
        }

        switch obj := revision.Data.Object.(type) {
        case *instancetypev1alpha2.VirtualMachinePreference:
                return &obj.Spec, nil
        case *instancetypev1alpha2.VirtualMachineClusterPreference:
                return &obj.Spec, nil
        default:
                return nil, fmt.Errorf("unexpected type in ControllerRevision: %T", obj)
        }
}
pkg/instancetype/instancetype.go:571: 571-599 lines are duplicate of `pkg/instancetype/instancetype.go:416-444` (dupl)
func (m *InstancetypeMethods) findInstancetypeSpecRevision(namespacedName types.NamespacedName) (*instancetypev1alpha2.VirtualMachineInstancetypeSpec, error) {
        var (
                err      error
                revision *appsv1.ControllerRevision
        )

        if m.ControllerRevisionStore != nil {
                revision, err = m.getControllerRevisionByInformer(namespacedName)
        } else {
                revision, err = m.getControllerRevisionByClient(namespacedName)
        }

        if err != nil {
                return nil, err
        }

        if err := decodeRevisionObject(revision, false); err != nil {
                return nil, err
        }

        switch obj := revision.Data.Object.(type) {
        case *instancetypev1alpha2.VirtualMachineInstancetype:
                return &obj.Spec, nil
        case *instancetypev1alpha2.VirtualMachineClusterInstancetype:
                return &obj.Spec, nil
        default:
                return nil, fmt.Errorf("unexpected type in ControllerRevision: %T", obj)
        }
}
pkg/instancetype/instancetype.go:844: 844-858 lines are duplicate of `pkg/instancetype/instancetype.go:860-874` (dupl)
func AddInstancetypeNameAnnotations(vm *virtv1.VirtualMachine, target metav1.Object) {
        if vm.Spec.Instancetype == nil {
                return
        }

        if target.GetAnnotations() == nil {
                target.SetAnnotations(make(map[string]string))
        }
        switch strings.ToLower(vm.Spec.Instancetype.Kind) {
        case apiinstancetype.PluralResourceName, apiinstancetype.SingularResourceName:
                target.GetAnnotations()[virtv1.InstancetypeAnnotation] = vm.Spec.Instancetype.Name
        case "", apiinstancetype.ClusterPluralResourceName, apiinstancetype.ClusterSingularResourceName:
                target.GetAnnotations()[virtv1.ClusterInstancetypeAnnotation] = vm.Spec.Instancetype.Name
        }
}
pkg/instancetype/instancetype.go:860: 860-874 lines are duplicate of `pkg/instancetype/instancetype.go:844-858` (dupl)
func AddPreferenceNameAnnotations(vm *virtv1.VirtualMachine, target metav1.Object) {
        if vm.Spec.Preference == nil {
                return
        }

        if target.GetAnnotations() == nil {
                target.SetAnnotations(make(map[string]string))
        }
        switch strings.ToLower(vm.Spec.Preference.Kind) {
        case apiinstancetype.PluralPreferenceResourceName, apiinstancetype.SingularPreferenceResourceName:
                target.GetAnnotations()[virtv1.PreferenceAnnotation] = vm.Spec.Preference.Name
        case "", apiinstancetype.ClusterPluralPreferenceResourceName, apiinstancetype.ClusterSingularPreferenceResourceName:
                target.GetAnnotations()[virtv1.ClusterPreferenceAnnotation] = vm.Spec.Preference.Name
        }
}
pkg/instancetype/compatibility.go:36: 36-64 lines are duplicate of `pkg/instancetype/compatibility.go:66-94` (dupl)
func decodeOldInstancetypeRevisionObject(data []byte) (*instancetypev1alpha1.VirtualMachineInstancetype, error) {
        revision := &instancetypev1alpha1.VirtualMachineInstancetypeSpecRevision{}
        err := json.Unmarshal(data, revision)
        if err != nil {
                // Failed to unmarshal, so the object is not the expected type
                return nil, nil
        }

        if revision.APIVersion != instancetypev1alpha1.SchemeGroupVersion.String() {
                return nil, nil
        }

        instancetypeSpec := &instancetypev1alpha1.VirtualMachineInstancetypeSpec{}
        err = json.Unmarshal(revision.Spec, instancetypeSpec)
        if err != nil {
                return nil, fmt.Errorf("failed to unmarshal json into v1alpha1.VirtualMachineInstancetypeSpec: %w", err)
        }

        return &instancetypev1alpha1.VirtualMachineInstancetype{
                TypeMeta: metav1.TypeMeta{
                        APIVersion: instancetypev1alpha1.SchemeGroupVersion.String(),
                        Kind:       "VirtualMachineInstancetype",
                },
                ObjectMeta: metav1.ObjectMeta{
                        Name: "old-version-of-instancetype-object",
                },
                Spec: *instancetypeSpec,
        }, nil
}
pkg/instancetype/compatibility.go:66: 66-94 lines are duplicate of `pkg/instancetype/compatibility.go:36-64` (dupl)
func decodeOldPreferenceRevisionObject(data []byte) (*instancetypev1alpha1.VirtualMachinePreference, error) {
        revision := &instancetypev1alpha1.VirtualMachinePreferenceSpecRevision{}
        err := json.Unmarshal(data, revision)
        if err != nil {
                // Failed to unmarshall, so the object is not the expected type
                return nil, nil
        }

        if revision.APIVersion != instancetypev1alpha1.SchemeGroupVersion.String() {
                return nil, nil
        }

        preferenceSpec := &instancetypev1alpha1.VirtualMachinePreferenceSpec{}
        err = json.Unmarshal(revision.Spec, preferenceSpec)
        if err != nil {
                return nil, fmt.Errorf("failed to unmarshall json into v1alpha1.VirtualMachinePreferenceSpec: %w", err)
        }

        return &instancetypev1alpha1.VirtualMachinePreference{
                TypeMeta: metav1.TypeMeta{
                        APIVersion: instancetypev1alpha1.SchemeGroupVersion.String(),
                        Kind:       "VirtualMachinePreference",
                },
                ObjectMeta: metav1.ObjectMeta{
                        Name: "old-version-of-preference-object",
                },
                Spec: *preferenceSpec,
        }, nil
}
pkg/instancetype/instancetype_test.go:429: 429-466 lines are duplicate of `pkg/instancetype/instancetype_test.go:759-796` (dupl)
        Context("Add instancetype name annotations", func() {
                const instancetypeName = "instancetype-name"

                BeforeEach(func() {
                        vm = kubecli.NewMinimalVM("testvm")
                        vm.Spec.Instancetype = &v1.InstancetypeMatcher{Name: instancetypeName}
                })

                It("should add instancetype name annotation", func() {
                        vm.Spec.Instancetype.Kind = apiinstancetype.SingularResourceName

                        meta := &metav1.ObjectMeta{}
                        instancetype.AddInstancetypeNameAnnotations(vm, meta)

                        Expect(meta.Annotations[v1.InstancetypeAnnotation]).To(Equal(instancetypeName))
                        Expect(meta.Annotations[v1.ClusterInstancetypeAnnotation]).To(Equal(""))
                })

                It("should add cluster instancetype name annotation", func() {
                        vm.Spec.Instancetype.Kind = apiinstancetype.ClusterSingularResourceName

                        meta := &metav1.ObjectMeta{}
                        instancetype.AddInstancetypeNameAnnotations(vm, meta)

                        Expect(meta.Annotations[v1.InstancetypeAnnotation]).To(Equal(""))
                        Expect(meta.Annotations[v1.ClusterInstancetypeAnnotation]).To(Equal(instancetypeName))
                })

                It("should add cluster name annotation, if instancetype.kind is empty", func() {
                        vm.Spec.Instancetype.Kind = ""

                        meta := &metav1.ObjectMeta{}
                        instancetype.AddInstancetypeNameAnnotations(vm, meta)

                        Expect(meta.Annotations[v1.InstancetypeAnnotation]).To(Equal(""))
                        Expect(meta.Annotations[v1.ClusterInstancetypeAnnotation]).To(Equal(instancetypeName))
                })
        })
pkg/instancetype/instancetype_test.go:759: 759-796 lines are duplicate of `pkg/instancetype/instancetype_test.go:429-466` (dupl)
        Context("Add preference name annotations", func() {
                const preferenceName = "preference-name"

                BeforeEach(func() {
                        vm = kubecli.NewMinimalVM("testvm")
                        vm.Spec.Preference = &v1.PreferenceMatcher{Name: preferenceName}
                })

                It("should add preference name annotation", func() {
                        vm.Spec.Preference.Kind = apiinstancetype.SingularPreferenceResourceName

                        meta := &metav1.ObjectMeta{}
                        instancetype.AddPreferenceNameAnnotations(vm, meta)

                        Expect(meta.Annotations[v1.PreferenceAnnotation]).To(Equal(preferenceName))
                        Expect(meta.Annotations[v1.ClusterPreferenceAnnotation]).To(Equal(""))
                })

                It("should add cluster preference name annotation", func() {
                        vm.Spec.Preference.Kind = apiinstancetype.ClusterSingularPreferenceResourceName

                        meta := &metav1.ObjectMeta{}
                        instancetype.AddPreferenceNameAnnotations(vm, meta)

                        Expect(meta.Annotations[v1.PreferenceAnnotation]).To(Equal(""))
                        Expect(meta.Annotations[v1.ClusterPreferenceAnnotation]).To(Equal(preferenceName))
                })

                It("should add cluster name annotation, if preference.kind is empty", func() {
                        vm.Spec.Preference.Kind = ""

                        meta := &metav1.ObjectMeta{}
                        instancetype.AddPreferenceNameAnnotations(vm, meta)

                        Expect(meta.Annotations[v1.PreferenceAnnotation]).To(Equal(""))
                        Expect(meta.Annotations[v1.ClusterPreferenceAnnotation]).To(Equal(preferenceName))
                })
        })
pkg/instancetype/instancetype_test.go:1373: 1373-1387 lines are duplicate of `pkg/instancetype/instancetype_test.go:1389-1403` (dupl)
                        It("should apply BIOS preferences full to VMI", func() {
                                preferenceSpec = &instancetypev1alpha2.VirtualMachinePreferenceSpec{
                                        Firmware: &instancetypev1alpha2.FirmwarePreferences{
                                                PreferredUseBios:       pointer.Bool(true),
                                                PreferredUseBiosSerial: pointer.Bool(true),
                                                PreferredUseEfi:        pointer.Bool(false),
                                                PreferredUseSecureBoot: pointer.Bool(false),
                                        },
                                }

                                conflicts := instancetypeMethods.ApplyToVmi(field, instancetypeSpec, preferenceSpec, &vmi.Spec)
                                Expect(conflicts).To(HaveLen(0))

                                Expect(*vmi.Spec.Domain.Firmware.Bootloader.BIOS.UseSerial).To(Equal(*preferenceSpec.Firmware.PreferredUseBiosSerial))
                        })
pkg/instancetype/instancetype_test.go:1389: 1389-1403 lines are duplicate of `pkg/instancetype/instancetype_test.go:1373-1387` (dupl)
                        It("should apply SecureBoot preferences full to VMI", func() {
                                preferenceSpec = &instancetypev1alpha2.VirtualMachinePreferenceSpec{
                                        Firmware: &instancetypev1alpha2.FirmwarePreferences{
                                                PreferredUseBios:       pointer.Bool(false),
                                                PreferredUseBiosSerial: pointer.Bool(false),
                                                PreferredUseEfi:        pointer.Bool(true),
                                                PreferredUseSecureBoot: pointer.Bool(true),
                                        },
                                }

                                conflicts := instancetypeMethods.ApplyToVmi(field, instancetypeSpec, preferenceSpec, &vmi.Spec)
                                Expect(conflicts).To(HaveLen(0))

                                Expect(*vmi.Spec.Domain.Firmware.Bootloader.EFI.SecureBoot).To(Equal(*preferenceSpec.Firmware.PreferredUseSecureBoot))
                        })
pkg/instancetype/instancetype_test.go:583: 583-611 lines are duplicate of `pkg/instancetype/instancetype_test.go:713-742` (dupl)
                        It("store returns nil revision when RevisionName already populated", func() {
                                clusterPreferenceControllerRevision, err := instancetype.CreateControllerRevision(vm, clusterPreference)
                                Expect(err).ToNot(HaveOccurred())

                                _, err = virtClient.AppsV1().ControllerRevisions(vm.Namespace).Create(context.Background(), clusterPreferenceControllerRevision, metav1.CreateOptions{})
                                Expect(err).ToNot(HaveOccurred())

                                vm.Spec.Preference.RevisionName = clusterPreferenceControllerRevision.Name

                                Expect(instancetypeMethods.StoreControllerRevisions(vm)).To(Succeed())
                                Expect(vm.Spec.Preference.RevisionName).To(Equal(clusterPreferenceControllerRevision.Name))
                        })

                        It("store ControllerRevision succeeds if a revision exists with expected data", func() {
                                clusterPreferenceControllerRevision, err := instancetype.CreateControllerRevision(vm, clusterPreference)
                                Expect(err).ToNot(HaveOccurred())

                                _, err = virtClient.AppsV1().ControllerRevisions(vm.Namespace).Create(context.Background(), clusterPreferenceControllerRevision, metav1.CreateOptions{})
                                Expect(err).ToNot(HaveOccurred())

                                expectedRevisionNamePatch, err := instancetype.GenerateRevisionNamePatch(nil, clusterPreferenceControllerRevision)
                                Expect(err).ToNot(HaveOccurred())

                                vmInterface.EXPECT().Patch(vm.Name, types.JSONPatchType, expectedRevisionNamePatch, &metav1.PatchOptions{})

                                Expect(instancetypeMethods.StoreControllerRevisions(vm)).To(Succeed())
                                Expect(vm.Spec.Preference.RevisionName).To(Equal(clusterPreferenceControllerRevision.Name))

                        })
pkg/instancetype/instancetype_test.go:713: 713-742 lines are duplicate of `pkg/instancetype/instancetype_test.go:583-611` (dupl)
                        It("store returns nil revision when RevisionName already populated", func() {
                                preferenceControllerRevision, err := instancetype.CreateControllerRevision(vm, preference)
                                Expect(err).ToNot(HaveOccurred())

                                _, err = virtClient.AppsV1().ControllerRevisions(vm.Namespace).Create(context.Background(), preferenceControllerRevision, metav1.CreateOptions{})
                                Expect(err).ToNot(HaveOccurred())

                                vm.Spec.Preference.RevisionName = preferenceControllerRevision.Name

                                Expect(instancetypeMethods.StoreControllerRevisions(vm)).To(Succeed())
                                Expect(vm.Spec.Preference.RevisionName).To(Equal(preferenceControllerRevision.Name))

                        })

                        It("store ControllerRevision succeeds if a revision exists with expected data", func() {
                                preferenceControllerRevision, err := instancetype.CreateControllerRevision(vm, preference)
                                Expect(err).ToNot(HaveOccurred())

                                _, err = virtClient.AppsV1().ControllerRevisions(vm.Namespace).Create(context.Background(), preferenceControllerRevision, metav1.CreateOptions{})
                                Expect(err).ToNot(HaveOccurred())

                                expectedRevisionNamePatch, err := instancetype.GenerateRevisionNamePatch(nil, preferenceControllerRevision)
                                Expect(err).ToNot(HaveOccurred())

                                vmInterface.EXPECT().Patch(vm.Name, types.JSONPatchType, expectedRevisionNamePatch, &metav1.PatchOptions{})

                                Expect(instancetypeMethods.StoreControllerRevisions(vm)).To(Succeed())
                                Expect(vm.Spec.Preference.RevisionName).To(Equal(preferenceControllerRevision.Name))

                        })
pkg/instancetype/instancetype_test.go:1076: 1076-1108 lines are duplicate of `pkg/instancetype/instancetype_test.go:1110-1142` (dupl)
                Context("instancetype.Spec.GPUs", func() {

                        BeforeEach(func() {
                                instancetypeSpec = &instancetypev1alpha2.VirtualMachineInstancetypeSpec{
                                        GPUs: []v1.GPU{
                                                {
                                                        Name:       "barfoo",
                                                        DeviceName: "vendor.com/gpu_name",
                                                },
                                        },
                                }
                        })

                        It("should apply to VMI", func() {
                                conflicts := instancetypeMethods.ApplyToVmi(field, instancetypeSpec, preferenceSpec, &vmi.Spec)
                                Expect(conflicts).To(HaveLen(0))

                                Expect(vmi.Spec.Domain.Devices.GPUs).To(Equal(instancetypeSpec.GPUs))
                        })

                        It("should detect GPU conflict", func() {
                                vmi.Spec.Domain.Devices.GPUs = []v1.GPU{
                                        {
                                                Name:       "foobar",
                                                DeviceName: "vendor.com/gpu_name",
                                        },
                                }

                                conflicts := instancetypeMethods.ApplyToVmi(field, instancetypeSpec, preferenceSpec, &vmi.Spec)
                                Expect(conflicts).To(HaveLen(1))
                                Expect(conflicts[0].String()).To(Equal("spec.template.spec.domain.devices.gpus"))
                        })
                })
pkg/instancetype/instancetype_test.go:1110: 1110-1142 lines are duplicate of `pkg/instancetype/instancetype_test.go:1076-1108` (dupl)
                Context("instancetype.Spec.HostDevices", func() {

                        BeforeEach(func() {
                                instancetypeSpec = &instancetypev1alpha2.VirtualMachineInstancetypeSpec{
                                        HostDevices: []v1.HostDevice{
                                                {
                                                        Name:       "foobar",
                                                        DeviceName: "vendor.com/device_name",
                                                },
                                        },
                                }
                        })

                        It("should apply to VMI", func() {
                                conflicts := instancetypeMethods.ApplyToVmi(field, instancetypeSpec, preferenceSpec, &vmi.Spec)
                                Expect(conflicts).To(HaveLen(0))

                                Expect(vmi.Spec.Domain.Devices.HostDevices).To(Equal(instancetypeSpec.HostDevices))
                        })

                        It("should detect HostDevice conflict", func() {
                                vmi.Spec.Domain.Devices.HostDevices = []v1.HostDevice{
                                        {
                                                Name:       "foobar",
                                                DeviceName: "vendor.com/device_name",
                                        },
                                }

                                conflicts := instancetypeMethods.ApplyToVmi(field, instancetypeSpec, preferenceSpec, &vmi.Spec)
                                Expect(conflicts).To(HaveLen(1))
                                Expect(conflicts[0].String()).To(Equal("spec.template.spec.domain.devices.hostDevices"))
                        })
                })
pkg/instancetype/instancetype_test.go:248: 248-262 lines are duplicate of `pkg/instancetype/instancetype_test.go:390-405` (dupl)
                        It("store ControllerRevision succeeds if a revision exists with expected data", func() {
                                instancetypeControllerRevision, err := instancetype.CreateControllerRevision(vm, clusterInstancetype)
                                Expect(err).ToNot(HaveOccurred())

                                _, err = virtClient.AppsV1().ControllerRevisions(vm.Namespace).Create(context.Background(), instancetypeControllerRevision, metav1.CreateOptions{})
                                Expect(err).ToNot(HaveOccurred())

                                expectedRevisionNamePatch, err := instancetype.GenerateRevisionNamePatch(instancetypeControllerRevision, nil)
                                Expect(err).ToNot(HaveOccurred())

                                vmInterface.EXPECT().Patch(vm.Name, types.JSONPatchType, expectedRevisionNamePatch, &metav1.PatchOptions{})

                                Expect(instancetypeMethods.StoreControllerRevisions(vm)).To(Succeed())
                                Expect(vm.Spec.Instancetype.RevisionName).To(Equal(instancetypeControllerRevision.Name))
                        })
pkg/instancetype/instancetype_test.go:390: 390-405 lines are duplicate of `pkg/instancetype/instancetype_test.go:596-611` (dupl)
                        It("store ControllerRevision succeeds if a revision exists with expected data", func() {
                                instancetypeControllerRevision, err := instancetype.CreateControllerRevision(vm, fakeInstancetype)
                                Expect(err).ToNot(HaveOccurred())

                                _, err = virtClient.AppsV1().ControllerRevisions(vm.Namespace).Create(context.Background(), instancetypeControllerRevision, metav1.CreateOptions{})
                                Expect(err).ToNot(HaveOccurred())

                                expectedRevisionNamePatch, err := instancetype.GenerateRevisionNamePatch(instancetypeControllerRevision, nil)
                                Expect(err).ToNot(HaveOccurred())

                                vmInterface.EXPECT().Patch(vm.Name, types.JSONPatchType, expectedRevisionNamePatch, &metav1.PatchOptions{})

                                Expect(instancetypeMethods.StoreControllerRevisions(vm)).To(Succeed())
                                Expect(vm.Spec.Instancetype.RevisionName).To(Equal(instancetypeControllerRevision.Name))

                        })
pkg/instancetype/instancetype_test.go:596: 596-611 lines are duplicate of `pkg/instancetype/instancetype_test.go:727-742` (dupl)
                        It("store ControllerRevision succeeds if a revision exists with expected data", func() {
                                clusterPreferenceControllerRevision, err := instancetype.CreateControllerRevision(vm, clusterPreference)
                                Expect(err).ToNot(HaveOccurred())

                                _, err = virtClient.AppsV1().ControllerRevisions(vm.Namespace).Create(context.Background(), clusterPreferenceControllerRevision, metav1.CreateOptions{})
                                Expect(err).ToNot(HaveOccurred())

                                expectedRevisionNamePatch, err := instancetype.GenerateRevisionNamePatch(nil, clusterPreferenceControllerRevision)
                                Expect(err).ToNot(HaveOccurred())

                                vmInterface.EXPECT().Patch(vm.Name, types.JSONPatchType, expectedRevisionNamePatch, &metav1.PatchOptions{})

                                Expect(instancetypeMethods.StoreControllerRevisions(vm)).To(Succeed())
                                Expect(vm.Spec.Preference.RevisionName).To(Equal(clusterPreferenceControllerRevision.Name))

                        })
pkg/instancetype/instancetype_test.go:727: 727-742 lines are duplicate of `pkg/instancetype/instancetype_test.go:248-262` (dupl)
                        It("store ControllerRevision succeeds if a revision exists with expected data", func() {
                                preferenceControllerRevision, err := instancetype.CreateControllerRevision(vm, preference)
                                Expect(err).ToNot(HaveOccurred())

                                _, err = virtClient.AppsV1().ControllerRevisions(vm.Namespace).Create(context.Background(), preferenceControllerRevision, metav1.CreateOptions{})
                                Expect(err).ToNot(HaveOccurred())

                                expectedRevisionNamePatch, err := instancetype.GenerateRevisionNamePatch(nil, preferenceControllerRevision)
                                Expect(err).ToNot(HaveOccurred())

                                vmInterface.EXPECT().Patch(vm.Name, types.JSONPatchType, expectedRevisionNamePatch, &metav1.PatchOptions{})

                                Expect(instancetypeMethods.StoreControllerRevisions(vm)).To(Succeed())
                                Expect(vm.Spec.Preference.RevisionName).To(Equal(preferenceControllerRevision.Name))

                        })
pkg/instancetype/instancetype_test.go:224: 224-239 lines are duplicate of `pkg/instancetype/instancetype_test.go:373-388` (dupl)
                        It("store returns a nil revision when RevisionName already populated", func() {
                                clusterInstancetypeControllerRevision, err := instancetype.CreateControllerRevision(vm, clusterInstancetype)
                                Expect(err).ToNot(HaveOccurred())

                                _, err = virtClient.AppsV1().ControllerRevisions(vm.Namespace).Create(context.Background(), clusterInstancetypeControllerRevision, metav1.CreateOptions{})
                                Expect(err).ToNot(HaveOccurred())

                                vm.Spec.Instancetype = &v1.InstancetypeMatcher{
                                        Name:         clusterInstancetype.Name,
                                        RevisionName: clusterInstancetypeControllerRevision.Name,
                                        Kind:         apiinstancetype.ClusterSingularResourceName,
                                }

                                Expect(instancetypeMethods.StoreControllerRevisions(vm)).To(Succeed())
                                Expect(vm.Spec.Instancetype.RevisionName).To(Equal(clusterInstancetypeControllerRevision.Name))
                        })
pkg/instancetype/instancetype_test.go:373: 373-388 lines are duplicate of `pkg/instancetype/instancetype_test.go:224-239` (dupl)
                        It("store returns a nil revision when RevisionName already populated", func() {
                                instancetypeControllerRevision, err := instancetype.CreateControllerRevision(vm, fakeInstancetype)
                                Expect(err).ToNot(HaveOccurred())

                                _, err = virtClient.AppsV1().ControllerRevisions(vm.Namespace).Create(context.Background(), instancetypeControllerRevision, metav1.CreateOptions{})
                                Expect(err).ToNot(HaveOccurred())

                                vm.Spec.Instancetype = &v1.InstancetypeMatcher{
                                        Name:         fakeInstancetype.Name,
                                        RevisionName: instancetypeControllerRevision.Name,
                                        Kind:         apiinstancetype.SingularResourceName,
                                }

                                Expect(instancetypeMethods.StoreControllerRevisions(vm)).To(Succeed())
                                Expect(vm.Spec.Instancetype.RevisionName).To(Equal(instancetypeControllerRevision.Name))
                        })
pkg/instancetype/compatibility_test.go:144: Function 'getOldPreferenceSpec' is too long (116 > 100) (funlen)
func getOldPreferenceSpec() v1alpha1.VirtualMachinePreferenceSpec {
pkg/instancetype/instancetype_test.go:202:33: string `non-existing-instancetype` has 4 occurrences, make it a constant (goconst)
                                vm.Spec.Instancetype.Name = "non-existing-instancetype"
                                                            ^
pkg/instancetype/instancetype_test.go:555:31: string `non-existing-preference` has 4 occurrences, make it a constant (goconst)
                                vm.Spec.Preference.Name = "non-existing-preference"
                                                          ^
pkg/instancetype/instancetype.go:261:2: importShadow: shadow of imported from 'kubevirt.io/kubevirt/pkg/apimachinery/patch' package 'patch' (gocritic)
        patch, err := GenerateRevisionNamePatch(instancetypeRevision, preferenceRevision)
        ^
pkg/instancetype/instancetype.go:795:2: singleCaseSwitch: should rewrite switch statement to if statement (gocritic)
        switch sourceRef.Kind {
        ^
pkg/instancetype/instancetype.go:117:6: sloppyReassign: re-assignment to `err` can be replaced with `err := m.checkForInstancetypeConflicts(&instancetype.Spec, &vm.Spec.Template.Spec)` (gocritic)
                if err = m.checkForInstancetypeConflicts(&instancetype.Spec, &vm.Spec.Template.Spec); err != nil {
                   ^
pkg/instancetype/instancetype.go:131:6: sloppyReassign: re-assignment to `err` can be replaced with `err := m.checkForInstancetypeConflicts(&clusterInstancetype.Spec, &vm.Spec.Template.Spec)` (gocritic)
                if err = m.checkForInstancetypeConflicts(&clusterInstancetype.Spec, &vm.Spec.Template.Spec); err != nil {
                   ^
pkg/instancetype/instancetype.go:432:5: sloppyReassign: re-assignment to `err` can be replaced with `err := decodeRevisionObject(revision, true)` (gocritic)
        if err = decodeRevisionObject(revision, true); err != nil {
           ^
pkg/instancetype/instancetype.go:711:1: unnamedResult: consider giving a name to these results (gocritic)
func (m *InstancetypeMethods) inferDefaultsFromVolumes(vm *virtv1.VirtualMachine, inferFromVolumeName, defaultNameLabel, defaultKindLabel string) (string, string, error) {
^
pkg/instancetype/instancetype.go:727:1: unnamedResult: consider giving a name to these results (gocritic)
func inferDefaultsFromLabels(labels map[string]string, defaultNameLabel, defaultKindLabel string) (string, string, error) {
^
pkg/instancetype/instancetype.go:735:1: unnamedResult: consider giving a name to these results (gocritic)
func (m *InstancetypeMethods) inferDefaultsFromPVC(pvcName, pvcNamespace, defaultNameLabel, defaultKindLabel string) (string, string, error) {
^
pkg/instancetype/instancetype.go:1132:1: cyclomatic complexity 30 of func `applyHyperVFeaturePreferences` is high (> 15) (gocyclo)
func applyHyperVFeaturePreferences(preferenceSpec *instancetypev1alpha2.VirtualMachinePreferenceSpec, vmiSpec *virtv1.VirtualMachineInstanceSpec) {
^
pkg/instancetype/instancetype.go:965:1: cyclomatic complexity 26 of func `ApplyDevicePreferences` is high (> 15) (gocyclo)
func ApplyDevicePreferences(preferenceSpec *instancetypev1alpha2.VirtualMachinePreferenceSpec, vmiSpec *virtv1.VirtualMachineInstanceSpec) {
^
pkg/instancetype/instancetype.go:1031:1: cyclomatic complexity 22 of func `applyDiskPreferences` is high (> 15) (gocyclo)
func applyDiskPreferences(preferenceSpec *instancetypev1alpha2.VirtualMachinePreferenceSpec, vmiSpec *virtv1.VirtualMachineInstanceSpec) {
^
pkg/instancetype/instancetype.go:1196:1: cyclomatic complexity 16 of func `applyFirmwarePreferences` is high (> 15) (gocyclo)
func applyFirmwarePreferences(preferenceSpec *instancetypev1alpha2.VirtualMachinePreferenceSpec, vmiSpec *virtv1.VirtualMachineInstanceSpec) {
^
pkg/instancetype/compatibility.go:99: line is 141 characters (lll)
func convertInstancetype(source *instancetypev1alpha1.VirtualMachineInstancetype) (*instancetypev1alpha2.VirtualMachineInstancetype, error) {
pkg/instancetype/instancetype.go:33: line is 224 characters (lll)
        ApplyToVmi(field *k8sfield.Path, instancetypespec *instancetypev1alpha2.VirtualMachineInstancetypeSpec, prefernceSpec *instancetypev1alpha2.VirtualMachinePreferenceSpec, vmiSpec *virtv1.VirtualMachineInstanceSpec) Conflicts
pkg/instancetype/instancetype.go:96: line is 182 characters (lll)
func (m *InstancetypeMethods) checkForInstancetypeConflicts(instancetypeSpec *instancetypev1alpha2.VirtualMachineInstancetypeSpec, vmiSpec *virtv1.VirtualMachineInstanceSpec) error {
pkg/instancetype/instancetype.go:297: line is 146 characters (lll)
func storeRevision(revision *appsv1.ControllerRevision, clientset kubecli.KubevirtClient, isPreference bool) (*appsv1.ControllerRevision, error) {
pkg/instancetype/instancetype.go:298: line is 146 characters (lll)
        createdRevision, err := clientset.AppsV1().ControllerRevisions(revision.Namespace).Create(context.Background(), revision, metav1.CreateOptions{})
pkg/instancetype/instancetype.go:305: line is 147 characters (lll)
                existingRevision, err := clientset.AppsV1().ControllerRevisions(revision.Namespace).Get(context.Background(), revision.Name, metav1.GetOptions{})
pkg/instancetype/instancetype.go:360: line is 256 characters (lll)
func (m *InstancetypeMethods) ApplyToVmi(field *k8sfield.Path, instancetypeSpec *instancetypev1alpha2.VirtualMachineInstancetypeSpec, preferenceSpec *instancetypev1alpha2.VirtualMachinePreferenceSpec, vmiSpec *virtv1.VirtualMachineInstanceSpec) Conflicts {
pkg/instancetype/instancetype.go:462: line is 152 characters (lll)
        revision, err := m.Clientset.AppsV1().ControllerRevisions(namespacedName.Namespace).Get(context.Background(), namespacedName.Name, metav1.GetOptions{})
pkg/instancetype/instancetype.go:483: line is 149 characters (lll)
func (m *InstancetypeMethods) findPreferenceByInformer(namespacedName types.NamespacedName) (*instancetypev1alpha2.VirtualMachinePreference, error) {
pkg/instancetype/instancetype.go:498: line is 147 characters (lll)
func (m *InstancetypeMethods) findPreferenceByClient(namespacedName types.NamespacedName) (*instancetypev1alpha2.VirtualMachinePreference, error) {
pkg/instancetype/instancetype.go:499: line is 150 characters (lll)
        preference, err := m.Clientset.VirtualMachinePreference(namespacedName.Namespace).Get(context.Background(), namespacedName.Name, metav1.GetOptions{})
pkg/instancetype/instancetype.go:506: line is 143 characters (lll)
func (m *InstancetypeMethods) findClusterPreference(vm *virtv1.VirtualMachine) (*instancetypev1alpha2.VirtualMachineClusterPreference, error) {
pkg/instancetype/instancetype.go:516: line is 147 characters (lll)
func (m *InstancetypeMethods) findClusterPreferenceByInformer(resourceName string) (*instancetypev1alpha2.VirtualMachineClusterPreference, error) {
pkg/instancetype/instancetype.go:531: line is 145 characters (lll)
func (m *InstancetypeMethods) findClusterPreferenceByClient(resourceName string) (*instancetypev1alpha2.VirtualMachineClusterPreference, error) {
pkg/instancetype/instancetype.go:539: line is 141 characters (lll)
func (m *InstancetypeMethods) FindInstancetypeSpec(vm *virtv1.VirtualMachine) (*instancetypev1alpha2.VirtualMachineInstancetypeSpec, error) {
pkg/instancetype/instancetype.go:615: line is 153 characters (lll)
func (m *InstancetypeMethods) findInstancetypeByInformer(namespacedName types.NamespacedName) (*instancetypev1alpha2.VirtualMachineInstancetype, error) {
pkg/instancetype/instancetype.go:630: line is 151 characters (lll)
func (m *InstancetypeMethods) findInstancetypeByClient(namespacedName types.NamespacedName) (*instancetypev1alpha2.VirtualMachineInstancetype, error) {
pkg/instancetype/instancetype.go:631: line is 154 characters (lll)
        instancetype, err := m.Clientset.VirtualMachineInstancetype(namespacedName.Namespace).Get(context.Background(), namespacedName.Name, metav1.GetOptions{})
pkg/instancetype/instancetype.go:648: line is 151 characters (lll)
func (m *InstancetypeMethods) findClusterInstancetypeByInformer(resourceName string) (*instancetypev1alpha2.VirtualMachineClusterInstancetype, error) {
pkg/instancetype/instancetype.go:663: line is 149 characters (lll)
func (m *InstancetypeMethods) findClusterInstancetypeByClient(resourceName string) (*instancetypev1alpha2.VirtualMachineClusterInstancetype, error) {
pkg/instancetype/instancetype.go:675: line is 190 characters (lll)
        defaultName, defaultKind, err := m.inferDefaultsFromVolumes(vm, vm.Spec.Instancetype.InferFromVolume, apiinstancetype.DefaultInstancetypeLabel, apiinstancetype.DefaultInstancetypeKindLabel)
pkg/instancetype/instancetype.go:689: line is 184 characters (lll)
        defaultName, defaultKind, err := m.inferDefaultsFromVolumes(vm, vm.Spec.Preference.InferFromVolume, apiinstancetype.DefaultPreferenceLabel, apiinstancetype.DefaultPreferenceKindLabel)
pkg/instancetype/instancetype.go:779: line is 144 characters (lll)
        ds, err := m.Clientset.CdiClient().CdiV1beta1().DataSources(dataSourceNamespace).Get(context.Background(), dataSourceName, metav1.GetOptions{})
pkg/instancetype/instancetype.go:802: line is 229 characters (lll)
func applyCpu(field *k8sfield.Path, instancetypeSpec *instancetypev1alpha2.VirtualMachineInstancetypeSpec, preferenceSpec *instancetypev1alpha2.VirtualMachinePreferenceSpec, vmiSpec *virtv1.VirtualMachineInstanceSpec) Conflicts {
pkg/instancetype/instancetype.go:876: line is 165 characters (lll)
func applyMemory(field *k8sfield.Path, instancetypeSpec *instancetypev1alpha2.VirtualMachineInstancetypeSpec, vmiSpec *virtv1.VirtualMachineInstanceSpec) Conflicts {
pkg/instancetype/instancetype.go:902: line is 173 characters (lll)
func applyIOThreadPolicy(field *k8sfield.Path, instancetypeSpec *instancetypev1alpha2.VirtualMachineInstancetypeSpec, vmiSpec *virtv1.VirtualMachineInstanceSpec) Conflicts {
pkg/instancetype/instancetype.go:918: line is 173 characters (lll)
func applyLaunchSecurity(field *k8sfield.Path, instancetypeSpec *instancetypev1alpha2.VirtualMachineInstancetypeSpec, vmiSpec *virtv1.VirtualMachineInstanceSpec) Conflicts {
pkg/instancetype/instancetype.go:933: line is 163 characters (lll)
func applyGPUs(field *k8sfield.Path, instancetypeSpec *instancetypev1alpha2.VirtualMachineInstancetypeSpec, vmiSpec *virtv1.VirtualMachineInstanceSpec) Conflicts {
pkg/instancetype/instancetype.go:949: line is 170 characters (lll)
func applyHostDevices(field *k8sfield.Path, instancetypeSpec *instancetypev1alpha2.VirtualMachineInstancetypeSpec, vmiSpec *virtv1.VirtualMachineInstanceSpec) Conflicts {
pkg/instancetype/instancetype.go:1073: line is 143 characters (lll)
func applyInterfacePreferences(preferenceSpec *instancetypev1alpha2.VirtualMachinePreferenceSpec, vmiSpec *virtv1.VirtualMachineInstanceSpec) {
pkg/instancetype/instancetype.go:1095: line is 141 characters (lll)
func applyFeaturePreferences(preferenceSpec *instancetypev1alpha2.VirtualMachinePreferenceSpec, vmiSpec *virtv1.VirtualMachineInstanceSpec) {
pkg/instancetype/instancetype.go:1210: line is 197 characters (lll)
        if preferenceSpec.Firmware.PreferredUseBios != nil && *preferenceSpec.Firmware.PreferredUseBios && vmiSpec.Domain.Firmware.Bootloader.BIOS == nil && vmiSpec.Domain.Firmware.Bootloader.EFI == nil {
pkg/instancetype/instancetype.go:1218: line is 195 characters (lll)
        if preferenceSpec.Firmware.PreferredUseEfi != nil && *preferenceSpec.Firmware.PreferredUseEfi && vmiSpec.Domain.Firmware.Bootloader.EFI == nil && vmiSpec.Domain.Firmware.Bootloader.BIOS == nil {
pkg/instancetype/instancetype.go:1251: line is 145 characters (lll)
        // We don't want to allow a partial overwrite here as that could lead to some unexpected behaviour for users so only replace when nothing is set
pkg/instancetype/instancetype.go:1252: line is 153 characters (lll)
        if preferenceSpec.Clock.PreferredClockOffset != nil && vmiSpec.Domain.Clock.ClockOffset.UTC == nil && vmiSpec.Domain.Clock.ClockOffset.Timezone == nil {
pkg/instancetype/instancetype_test.go:228: line is 158 characters (lll)
                                _, err = virtClient.AppsV1().ControllerRevisions(vm.Namespace).Create(context.Background(), clusterInstancetypeControllerRevision, metav1.CreateOptions{})
pkg/instancetype/instancetype_test.go:252: line is 151 characters (lll)
                                _, err = virtClient.AppsV1().ControllerRevisions(vm.Namespace).Create(context.Background(), instancetypeControllerRevision, metav1.CreateOptions{})
pkg/instancetype/instancetype_test.go:281: line is 142 characters (lll)
                                Expect(instancetypeMethods.StoreControllerRevisions(vm)).To(MatchError(ContainSubstring("VM field conflicts with selected Instancetype")))
pkg/instancetype/instancetype_test.go:424: line is 142 characters (lll)
                                Expect(instancetypeMethods.StoreControllerRevisions(vm)).To(MatchError(ContainSubstring("VM field conflicts with selected Instancetype")))
pkg/instancetype/instancetype_test.go:587: line is 156 characters (lll)
                                _, err = virtClient.AppsV1().ControllerRevisions(vm.Namespace).Create(context.Background(), clusterPreferenceControllerRevision, metav1.CreateOptions{})
pkg/instancetype/instancetype_test.go:600: line is 156 characters (lll)
                                _, err = virtClient.AppsV1().ControllerRevisions(vm.Namespace).Create(context.Background(), clusterPreferenceControllerRevision, metav1.CreateOptions{})
pkg/instancetype/instancetype_test.go:620: line is 156 characters (lll)
                                _, err = virtClient.AppsV1().ControllerRevisions(vm.Namespace).Create(context.Background(), clusterPreferenceControllerRevision, metav1.CreateOptions{})
pkg/instancetype/instancetype_test.go:717: line is 149 characters (lll)
                                _, err = virtClient.AppsV1().ControllerRevisions(vm.Namespace).Create(context.Background(), preferenceControllerRevision, metav1.CreateOptions{})
pkg/instancetype/compatibility.go:120:80: convertPreference - result 0 (*kubevirt.io/api/instancetype/v1alpha2.VirtualMachinePreference) is always nil (unparam)
func convertPreference(source *instancetypev1alpha1.VirtualMachinePreference) (*instancetypev1alpha2.VirtualMachinePreference, error) {
                                                                               ^
pkg/instancetype/instancetype.go:244: unnecessary leading newline (whitespace)
func (m *InstancetypeMethods) StoreControllerRevisions(vm *virtv1.VirtualMachine) error {

pkg/instancetype/instancetype.go:1028: unnecessary trailing newline (whitespace)

}
pkg/instancetype/instancetype.go:1059: unnecessary trailing newline (whitespace)

                } else if vmiDisk.DiskDevice.CDRom != nil {
pkg/instancetype/instancetype.go:1064: unnecessary trailing newline (whitespace)

                } else if vmiDisk.DiskDevice.LUN != nil {
pkg/instancetype/instancetype.go:345:6: ST1003: func getInstancetypeApiSpec should be getInstancetypeAPISpec (stylecheck)
func getInstancetypeApiSpec(obj runtime.Object) (interface{}, error) {
     ^
```

lyarwood commented:
/unhold

I'll use a new PR for the lint feedback.

vladikr commented:
/approve
Looks good.

lyarwood commented:
/retest-required

0xFelix reviewed (approved):
/lgtm

0xFelix commented:
/approve
